wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFactoryBuilder.java

```diff
@@ -84,20 +91,28 @@ public URLFactoryBuilder(
    * Build a factory using the provided devfile
    *
    * @param devfileLocation location of devfile
+   * @param fileUrlComposer optional service-specific composer of URL's to the file raw content
    * @return a factory or null if devfile is not found
    */
-  public Optional<FactoryDto> createFactoryFromDevfile(String devfileLocation)
+  public Optional<FactoryDto> createFactoryFromDevfile(
+      String devfileLocation, @Nullable Function<String, String> fileUrlComposer)
       throws BadRequestException, ServerException {
     if (devfileLocation == null) {
       return Optional.empty();
     }
     final String devfileYamlContent = urlFetcher.fetch(devfileLocation);
-    if (Strings.isNullOrEmpty(devfileYamlContent)) {
+    if (isNullOrEmpty(devfileYamlContent)) {
       return Optional.empty();
     }
     try {
       Devfile devfile = devfileManager.parse(devfileYamlContent, false);
+      Optional<EnvironmentImpl> environment =
+          devfileEnvironmentProvisioner.tryProvision(devfile, fileUrlComposer);
       WorkspaceConfigImpl wsConfig = devfileManager.createWorkspaceConfig(devfile);
+      if (environment.isPresent()) {
+        wsConfig.setDefaultEnv("default");
```

Comment:
I think it would be better to use tool name as environment name instead of using `default` value. WDYT?

Reply:
Yes i thought abt this but it is little bit tricky since we have no tool itself here. I will try again.